### PR TITLE
Fix loading of modplug files

### PR DESCRIPTION
**Pull request by snue**
The settings for Modplug decoding are not restored/enforced on Mixxx startup since https://github.com/mixxxdj/mixxx/pull/3446. With the "default" maximum buffer size initialized to zero, all loading attempts of tracker files are doomed to fail. A round-trip to the settings dialog and hitting the "apply" button on the "Modplug Decoder" page works around the issue currently.

This Pull Request:
* Loads and applies modplug configuration when the preferences are created.
* Marks `SoundSourceModPlug` as `final` to silence a warning about virtual dispatch not working in the destructor.

### CODE REVIEW

**Comment by ronso0** [comment]
I never used modplug, though I wonder how such a bug could have slipped through during review.

> The settings for Modplug decoding are currently neither persisted nor restored/enforced on Mixxx startup

The settings are persistent, though they are appearantly not reloaded until the preferences are opened (calling `slotUpdate` for each pref page).
The fix is as easy as adding
```
loadSettings();
applySettings();
```
below this line https://github.com/mixxxdj/mixxx/blob/38059484217db1a425484875024ff48579367210/src/preferences/dialog/dlgprefmodplug.cpp#L42 so they are restored when the preferences are constructed.

If someone else can confirm the bug, this small fix should go to the 2.3 branch so it goes to 2.3.3
Even though it's just the buffer size here, defaults that are needed in various places could go to `src/sources/modplug_defs.h` for example?

**Comment by snue** [comment]
> The settings are persistent, though they are appearantly not reloaded until the preferences are opened (calling `slotUpdate` for each pref page).
> The fix is as easy as adding

Ah cool! I totally forgot how this works :)
I verified that fixes the issue for me. Will update the PR and try to change the target branch to 2.3.

**Comment by snue** [comment]
> If someone else can confirm the bug

@ronso0 if you are just short on tracker files you could replicate e.g. with [the one released here](https://www.infect-rules.de/2019/12/27/beathoven-infect-back-on-track-module-released-xmas-2019/).

**Comment by ronso0** [comment]
I can not confirm the bug, that file is loaded and played fine.
mixxx.cpp takes care of restoring the configuration
https://github.com/mixxxdj/mixxx/blob/38059484217db1a425484875024ff48579367210/src/mixxx.cpp#L393-L399
How come this is not working for you?

**Comment by snue** [comment]
> How come this is not working for you?

Interesting. I could reproduce on `2.3` and `main` and the proposed patch fixed it. I'll check again and report back.

**Comment by snue** [comment]
I retried this and indeed `2.3` is actually good. I can still reproduce the issue on `main` though. The `mixxx.cpp` is gone/moved to `mixxxmainwindow.cpp` there and initialization changed. Looks like that initialization code was dropped accidentally [when introducing `CoreServices`](https://github.com/mixxxdj/mixxx/commit/f754682adf2874041e0eddff708cfd396c805e72) in https://github.com/mixxxdj/mixxx/pull/3446.

Please let me know how you'd like to proceed. I'll switch the target branch back to `main` as `2.3` doesn't need fixing. But maybe we want to do initialization like before again instead of utilizing the constructor?

**Comment by ronso0** [comment]
> But maybe we want to do initialization like before again instead of utilizing the constructor?

Sure. I'm not familiar with coreservices, though. Just give it a try!

After this line I guess https://github.com/mixxxdj/mixxx/blob/6f8676073e1689a1cc749a701d909570d0cafb0e/src/coreservices.cpp#L314
``` c++
#ifdef __MODPLUG__
    // restore the configuration for the modplug library before trying to load a module
    DlgPrefModplug* pModplugPrefs = new DlgPrefModplug(0,
            m_pCoreServices->getSettingsManager().get());
    pModplugPrefs->loadSettings();
    pModplugPrefs->applySettings();
    delete pModplugPrefs; // not needed anymore
#endif
```

**Comment by ronso0** [comment]
_ping_ @snue 
Do you still intend to implement the proposed coreservices solution?

**Comment by snue** [comment]
Sure. I didn't find the time recently and figured it's not that pressing as 2.3 isn't affected. I'll try this out and update in the coming days hopefully. Stay tuned.

**Comment by ronso0** [comment]
sure, I didn't mean to push you ; )

**Comment by snue** [comment]
@ronso0 I finally got around to fix this. This is the "just make it work like before" approach. CoreServices should ideally not rely on any UI widget. But I think for now this temporary preferences widget is better than broken Modplug file support.

**Comment by ronso0** [comment]
Thank you! 

Regarding the 2nd commit, I'm not sufficiently familiar with the soundsource stuff.
@uklotzde or @Swiftb0y Do you mind taking a look?

**Comment by Swiftb0y** [comment]
Making things final should not do any harm if it still builds.

File: src/coreservices.cpp [hunk 1]
```diff
@@ -313,6 +316,14 @@ void CoreServices::initialize(QApplication* pApp) {
     m_pVCManager->init();
 #endif
 
+#ifdef __MODPLUG__
+    // Restore the configuration for the modplug library before trying to load a module.
+    DlgPrefModplug* pModplugPrefs = new DlgPrefModplug(nullptr, pConfig);
+    pModplugPrefs->loadSettings();
+    pModplugPrefs->applySettings();
+    delete pModplugPrefs; // not needed anymore
```
[the remark this answers is not in the document]
Comment (snue) [reply src/coreservices.cpp]: Yes that works as well. I switched to just allocate the temporary widget on the stack.

**Comment by Swiftb0y** [comment]
Please fix the code-style issue. then LGTM

**Review by Swiftb0y — APPROVED** [review]
Thanks. LGTM

**Comment by ronso0** [comment]
Thank you!

Unfortunately I can't test atm with the [file you shared](https://github.com/mixxxdj/mixxx/pull/4736#issuecomment-1114296629) due to https://bugs.launchpad.net/mixxx/+bug/1979864

**Comment by snue** [comment]
> Thank you!
> 
> Unfortunately I can't test atm with the [file you shared](https://github.com/mixxxdj/mixxx/pull/4736#issuecomment-1114296629) due to https://bugs.launchpad.net/mixxx/+bug/1979864

Oh, that's an interesting bug. But it's not modplug specific, right? Could happen with other sound files as well?

Regarding modplug files, https://modarchive.org has a wide collection of compatible tracks for download. Of course lots of cheesy music there as well. [This one](https://modarchive.org/index.php?request=view_by_moduleid&query=115668) should not kill your ears :stuck_out_tongue_winking_eye:

**Comment by ronso0** [comment]
Alright, I worked around the bug and the mod track plays fine.
Thank you for the fix!

**Comment by ronso0** [comment]
@snue Do you have interest and time to look into an alternative way to initialize the decoder, independet from DlgPrefModplug?

**Comment by snue** [comment]
> @snue Do you have interest and time to look into an alternative way to initialize the decoder, independet from DlgPrefModplug?

At least not immediately. I might come up with something in the future but won't complain if someone else does it first.

**Pull request merged**